### PR TITLE
Add swiftc type check times

### PR DESCRIPTION
**Pull request by ecamacho**
Parses the `swiftc` type check times generated when the compiler option `-Xfrontend -debug-time-expression-type-checking` is used. 
Shows the number of occurrences of each swift function compilation and swift type check. With that, you can calculate the cumulative time it actually took `swiftc` to compile your expression, as stated in this [blog post](https://medium.com/swift-programming/swift-build-time-optimizations-part-2-37b0a7514cbe).

Updates the HTML reporter to show the new data:

image

### CODE REVIEW

File: docs/JSON Format.md [hunk 1]
```diff
@@ -52,7 +70,8 @@ Other fields:
 - `warnings`: the list of warnings
 - `errorCount`: the number of errors for the given step.
 - `errors`: the list of errors.
-- `swiftFunctionTimes`: Optional. If the step is a `swiftCompilation` and the app was compiled with the flags `-Xfrontend -debug-time-function-bodies` it will show the list of functions and their compilation time.
+- `swiftFunctionTimes`: Optional. If the step is a `swiftCompilation` and the app was compiled with the flags `-Xfrontend -debug-time-function-bodies` it will show the list of functions and their compilation time. The `occurrences` is the number of times that function is compiled , so the effective time it took in your build is durationMS * ocurrences.
```
Comment (BalestraPatrick) [inline docs/JSON Format.md:73]: nit: can remove the space before the comma

File: Sources/XCLogParser/parser/ParserBuildSteps.swift [hunk 2]
```diff
@@ -267,34 +267,40 @@ public final class ParserBuildSteps {
                 "notes": notices.getNotes()]
     }
 
-    private func decorateWithFunctionSteps(_ mainStep: BuildStep) -> BuildStep {
-        swiftFunctionTimesParser.parse()
-        guard swiftFunctionTimesParser.hasFunctionTimes() else {
+    private func decorateWithSwiftCTimes(_ mainStep: BuildStep) -> BuildStep {
```
Comment (BalestraPatrick) [inline Sources/XCLogParser/parser/ParserBuildSteps.swift:270]: In some other places we spell it `Swiftc`, but here it's `SwiftC`. Should we choose one?

File: Sources/XCLogParser/parser/SwiftCompilerFunctionTimeOptionParser.swift [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        guard let regexp = regexp else {
+            return nil
+        }
+        let range = NSRange(location: 0, length: command.count)
+        let matches = regexp.matches(in: command, options: .reportProgress, range: range)
+        let functionTimes = matches.compactMap { result -> SwiftFunctionTime? in
+            let durationString = command.substring(result.range(at: 1))
+            let file = command.substring(result.range(at: 2))
+            // some entries are invalid, we discard them
+            if isInvalid(fileName: file) {
```
Comment (BalestraPatrick) [inline Sources/XCLogParser/parser/SwiftCompilerFunctionTimeOptionParser.swift:65]: In which cases are the entries invalid?

Comment (BalestraPatrick) [reply Sources/XCLogParser/parser/SwiftCompilerFunctionTimeOptionParser.swift:65]: Oh, it's just `swiftc` that can't figure out the location?

Comment (ecamacho) [reply Sources/XCLogParser/parser/SwiftCompilerFunctionTimeOptionParser.swift:65]: yes, so we don't have a way to know the file nor the line number

File: Sources/XCLogParser/parser/ParserBuildSteps.swift (hunk 2 above)
Comment (ecamacho) [reply Sources/XCLogParser/parser/ParserBuildSteps.swift]: Oh yeah it should be Swiftc, changing it

File: docs/JSON Format.md (hunk 1 above)
Comment (ecamacho) [reply docs/JSON Format.md]: 👍 done

**Pull request merged**